### PR TITLE
Improved router handlers architecture in Coach - part 1

### LINKED ISSUE #11219: Improve router handlers architecture in Coach
## Blocks

https://github.com/learningequality/kolibri/issues/11422

## Observed behavior

For majority of Coach pages, in addition to fetching data in their individual route handlers (e.g. [`showExamsPage` handler](https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js#L6)), we fetch some of their data in the global `beforeRoute` handler, which is run before individual route handlers:

https://github.com/learningequality/kolibri/blob/0f7b06cc320476cbe0aca4ecaacae107b184cd45/kolibri/plugins/coach/assets/src/app.js#L38-L57

This approach has the following problems:

**1. Slows down navigation (user-facing)**
Particularly observable on slower networks when after clicking a coach page, until all global `beforeRoute` promises are finished, user is not even redirected to a new page.

coach-loading

**2. Limits individual pages ability to control when and how to display loading states (user-facing)**
Many pages have their individual handlers that are sometimes connected to local loading states, however there is no way to factor in waiting time for this global `beforeRoute` which often seem to take even more time than promises in individual handlers, causing inaccuracy and discrepancy of loading states. 

**3. It's difficult to build a mental picture of what requests are performed when a page is visited (developer-facing)**
Promises in the global `beforeRoute` are run for majority of pages except of few pages listed here

https://github.com/learningequality/kolibri/blob/0f7b06cc320476cbe0aca4ecaacae107b184cd45/kolibri/plugins/coach/assets/src/app.js#L40

Then, many pages have their separate handlers that have no information or control over the common logic, and are placed in different files.

## Expected behavior

- Users are redirected immediately after navigating to another page and appropriate loaders are displayed on the page
- Individual pages have information on what data is being fetched, when it starts, ends and can control loading states in a flexible and simple manner

## Steps to reproduce

Navigate Coach main pages and tabs on slower network

## Guidance

To be able to achieve the expected behavior, we need to move promises from the global `beforeRoute` handler

https://github.com/learningequality/kolibri/blob/0f7b06cc320476cbe0aca4ecaacae107b184cd45/kolibri/plugins/coach/assets/src/app.js#L26 

to individual page handlers.

For example, for [`showExamsPage` page handler](https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js#L6) the steps could be:

1. Search the codebase to locate the name of a page that uses `showExamsPage`: `PageNames.EXAMS`
    
https://github.com/learningequality/kolibri/blob/b88a8043fd4a0071d281699ed853c791a55518d8/kolibri/plugins/coach/assets/src/routes/planExamRoutes.js#L27-L37

2. Add a condition to the global `beforeRoute` handler instructing it to not run code containing promises fetching data if user is navigating to a page with name `PageNames.EXAMS`

3. And instead, copy all those promises to `showExamsPage`

- This needs to be done for all individual page handlers in Coach. In this manner, it can be done gradually.  I would recommend opening one pull request for one handler so we can review and QA loading states easily.

- After all individual handlers are updated to have the promises, we can finally remove the promises completely from the global `beforeRoute` handlers.

## Acceptance criteria

- [ ] The global `beforeRoute` handler doesn't contain any promises fetching data
- [ ] The promises are instead in individual page handlers
- [ ] No regressions in regards to fetching data for Coach pages
- [ ] No regressions in the existing loading states

### Tasks
- [x] Refactor routes in _kolibri/plugins/coach/assets/src/routes/planLessonsRoutes.js_ (DONE in https://github.com/learningequality/kolibri/pull/11675)
- [x] Refactor routes in _kolibri/plugins/coach/assets/src/routes/planRoutes.js_ (DONE in https://github.com/learningequality/kolibri/pull/11900)
- [ ] Refactor routes in _kolibri/plugins/coach/assets/src/routes/index.js_ (WIP, first related PR merged https://github.com/learningequality/kolibri/pull/12358)
- [ ] Refactor routes in _kolibri/plugins/coach/assets/src/routes/reportRoutes.js_
- [ ] BLOCKED (as of February 2) Refactor routes in _kolibri/plugins/coach/assets/src/routes/planExamRoutes.js_
- [ ] Finally remove data fetching from the global `beforeRoute` handler + cleanup temporary ignore list from there as well

### PR DESCRIPTION
**Pull request by ShivangRawat30**
## Summary

This pull request enhances route handling by restricting the responsibility of route handling to local handlers rather than relying on a global route handler.

## References

This work is part of #11219 (it doesn't close it though)

## Reviewer guidance

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by marcellamaki** [comment]
Added @nucleogenesis as a reviewer just because I think there are some concurrent changes to coach routing in develop for Enhanced Quiz Management

File: kolibri/plugins/coach/assets/src/app.js
Comment (MisRob) [inline kolibri/plugins/coach/assets/src/app.js]: This `else` part can be removed since it is included in the section above. Just to make that new separation of logic dealing with fetching data from other logic clearer.

File: kolibri/plugins/coach/assets/src/app.js
Comment (MisRob) [inline kolibri/plugins/coach/assets/src/app.js]: I know I did this in my draft, I wasn't very diligent :)! I'd suggest we improve readability a bit by doing something like:

```
// temporary condition as we're gradually moving all promises below this line to local page handlers and therefore need to skip those that we already refactored here https://github.com/learningequality/kolibri/issues/11219
if (to.name && to.name === PageNames.EXAMS) {
  next()
}
if (
  to.name &&
  to.params.classId &&
  !['CoachClassListPage', 'StatusTestPage', 'CoachPrompts', 'AllFacilitiesPage'].includes(
    to.name
  )
)
etc etc.
```

Rather than wrapping all promises in another condition.

Also notice I added a comment so that we know what's that about just in case it would take us some time to refactor all handlers.

File: kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js
Comment (MisRob) [inline kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js:1]: Nice! I appreciate that you didn't copy-pasted the whole piece we'll be removing from the global route hook, but also cleaned up promises that weren't relevant to this page. This is a good approach to adopt in other handlers as well.

**Review by MisRob — COMMENTED** [review]
Hi @ShivangRawat30, in the light of we discussed, this makes sense to me. Thank you for following the suggestion to refactor the global beforeRoute hook into two separate sections - the first with other logic that will stay, and the latter with what will be eventually removed. I am leaving a few comments related to code. I will also test it for regressions.

I will update the PR description since this doesn't close the whole https://github.com/learningequality/kolibri/pull/11570. It's rather a first step. There are plenty of other page handlers that we need to add to the ignore list we now have thanks to this work in the global beforeRoute, and move promises to those page handlers. And after that's done, we will remove the second section of the global beforeRoute finally. Then, the issue will be closed. Would you like to continue working on it? The approach will be the same.

**Comment by MisRob** [comment]
@nucleogenesis this approach is based on https://github.com/learningequality/kolibri/issues/11219#issuecomment-1831913746, https://github.com/learningequality/kolibri/issues/11219#issuecomment-1832430052, and https://gist.github.com/MisRob/fe3d75e27883308272089dcfe02356ee to help @ShivangRawat30 to start this work gradually (just to make sure we're on the same board before refactoring all handlers). If he decides to continue, we will talk about in how large chunks this will get addressed.

If you haven't changed the core workings of the global beforeRoute hook, I don't think there should be many significant conflicts, right? Perhaps some in individual page handlers, but that's something that shouldn't be difficult to deal with. Please let me know how things are in the Enhanced Quiz Management.

**Comment by nucleogenesis** [comment]
The approach looks good but it does seem likely that the changes should all happen in one PR (I'm assuming the changes in the router guard will cause regressions in handlers.js's that haven't had the same update as examsRoot?).

In any case -- we are no longer using the handlers for exam creation in the routes and I'm not sure if the changes would need to be applied [there](https://github.com/learningequality/kolibri/blob/170b74f4f5b55497355b1c8fd5262bb19f1709dd/kolibri/plugins/coach/assets/src/modules/examCreation/handlers.js#L16) anyway.

(thanks for your work on this @ShivangRawat30 ! <3)

**Comment by ShivangRawat30** [comment]
@MisRob I've implemented the suggested changes for the review. and I am eager to continue working on this issue.

**Comment by MisRob** [comment]
@nucleogenesis

> The approach looks good but it does seem likely that the changes should all happen in one PR (I'm assuming the changes in the router guard will cause regressions in handlers.js's that haven't had the same update as examsRoot?).

No, the main idea here is actually to prepare ground for rather gradual refactor (for more reasons). Here, you can see that the promises we want to eventually remove are now ignored only for a page in which the first handler that Shivang refactored, while still being run for all other pages. We will keep adding to this "ignore" while refactoring other handlers. After all handlers are done, the ignore condition will be removed together with all promises fetching data from the global router handler:

https://github.com/learningequality/kolibri/pull/11570/files#diff-bf0e4e3732211307b4c4c45a916359f5a127f4065157500405c7a5411983a857R50-R52

```
     // temporary condition as we're gradually moving all promises below this line to local page handlers and therefore need to skip those that we already refactored here https://github.com/learningequality/kolibri/issues/11219
      if (to.name && to.name === PageNames.EXAMS) {
        next();
      }
```

The rest of this PR is only reordering the global handler to allow for a cleaner separation and a bit more readable structure during this temporary phase.

**Comment by MisRob** [comment]
@nucleogenesis

> In any case -- we are no longer using the handlers for exam creation in the routes and I'm not sure if the changes would need to be applied [there](https://github.com/learningequality/kolibri/blob/170b74f4f5b55497355b1c8fd5262bb19f1709dd/kolibri/plugins/coach/assets/src/modules/examCreation/handlers.js#L16) anyway.

This is a bigger concern. If you still rely on the global handler (I assume you do?) then I would say that probably you rely on those data being fetched in the global router being available and used somehow. So I think it's be best to left `examCreation` module untouched and leave as last part of this refactor after we agree together it's time, right? Due to the approach I explained above everything should still work just fine for all other handlers we didn't explicitly refactored, so I'd be okay with this.

**Comment by MisRob** [comment]
@nucleogenesis if this plan of attack makes sense to you too, is `examCreation` module the only module that's under work, or are there some more that'd be best to leave for later to prevent conflicts?

**Comment by nucleogenesis** [comment]
Only the examCreation :)

**Comment by MisRob** [comment]
@ShivangRawat30 Thank you! I will re-review and also test in detail that the global `beforeRoute` still works just the same for all other pages except the one you refactored. Then we will have a solid ground set up and it should be pretty easy now. I think then you could continue working pretty independently on plenty of other handlers. But based on what we've just talked with @nucleogenesis, I will take some time to plan a bit which handlers are good to tackle and provide you with a list.

**Comment by MisRob** [comment]
@nucleogenesis 

> Only the examCreation :)

Good news! Alright, thank you, that's good to know in advance.

File: kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js
Comment (MisRob) [inline kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js]: We need to make sure that the promises you moved into this handler will be run and resolved before all other promises that were present in this handler before. This is because some of the existing promises rely on the presence of data obtained by calling `'initClassInfo'` and `'getFacilities'` (in the previous architecture, this was implicitly fulfilled by having them placed in the global `beforeRoute` which was resolved before this page handler was run).

This should fix this regression error:

Screenshot from 2023-12-04 14-52-53

After making any changes I'd recommend checking the console errors and warnings. Sometimes it's possible there were 
some errors before, but there should be no new errors when compared to the latest `develop`. In this way, you can notice issues like the one I've just mentioned.

File: kolibri/plugins/coach/assets/src/app.js
Comment (MisRob) [inline kolibri/plugins/coach/assets/src/app.js:51]: I believe we also need to add `return;` below `next()` to ensure that code execution is stopped

File: kolibri/plugins/coach/assets/src/app.js
Comment (MisRob) [inline kolibri/plugins/coach/assets/src/app.js]: This condition is already present on line 56 so you can remove this second nested `if` block

**Review by MisRob — CHANGES_REQUESTED** [review]
Good progress @ShivangRawat30, thank you. I re-reviewed and am leaving some new notes.

**Comment by ShivangRawat30** [comment]
@MisRob I've implemented the required modifications, kindly review them. Once this pull request receives approval and is merged, I'll proceed to work on additional handlers.

File: kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js
Comment (MisRob) [inline kolibri/plugins/coach/assets/src/modules/examsRoot/handlers.js]: @ShivangRawat30 Placing the promise here doesn't ensure it is resolved before other promises (even though it may look so randomly)

An example:
```
const promise1 = new Promise(resolve => {
  setTimeout(() => {
    console.log("Promise 1 resolved");
    resolve();
  }, 2000)
})
  
const promise2 = new Promise(resolve => {
  setTimeout(() => {
    console.log("Promise 2 resolved");
    resolve();
  }, 500)
})

Promise.all([promise1, promise2])
```

Here you can see that the second promise is actually resolved before the first one.

We need to ensure the following:

- `'initClassInfo'` and `'getFacilities'` are run first (their order doesn't matter)
- after the two above are resolved, then the remaining promises are run

You will need to utilize `.then()` or `async/await`

**Comment by ShivangRawat30** [comment]
@MisRob Apologies for earlier. I've now made the adjustments. Please review and let me know if the approach need some changes.

**Review by MisRob — APPROVED** [review]
Thank you, @ShivangRawat30. It's looking good to me. I also manually tested the page and haven't observed any issues.

Just for your information, I pushed a commit disabling Jest tests for this handler. I will describe what's the problem with updating the current test suite here https://github.com/learningequality/kolibri/issues/11615 later on.

**Comment by MisRob** [comment]
As I pushed my own update, I'd welcome a second pair of eyes on this PR before merging. cc @rtibbles

**Review by rtibbles — APPROVED** [review]
Changes make sense - we can gradually move the global behaviour for each handler out of this!

**Pull request merged**
